shiny/_typing_extensions.py

```diff
@@ -0,0 +1,41 @@
+import sys
+
+if sys.version_info >= (3, 8):
+    from typing import Literal, Protocol, runtime_checkable
+else:
+    from typing_extensions import Literal, Protocol, runtime_checkable
+
+if sys.version_info >= (3, 10):
+    from typing import TypeGuard
+else:
+    from typing_extensions import TypeGuard
+
+# Even though TypedDict is available in Python 3.8, because it's used with NotRequired,
+# they should both come from the same typing module.
+# https://peps.python.org/pep-0655/#usage-in-python-3-11
+if sys.version_info >= (3, 11):
+    from typing import NotRequired, TypedDict
+else:
+    from typing_extensions import NotRequired, TypedDict
```

Comment:
All of the pyright problems stem from `NotRequired` not being recognized with `TypedDict`. If I perform this import statement within the destination file (like the original code), it works as expected.

@wch Any idea why this is failing? It'd be nice to not have the version checking throughout the repo and consolidate it all into a single file.

Reply:
I think the problem may actually be totally different -- the main branch of htmltools has updated types, which would cause failures in CI here. PR #416 in this repo makes use of the new types.

What's surprising to me is that in GHA, it's simply skipping the "Type check with pyright" step on Python 3.10 and 3.11. Now that I look at it, I think it's because in the workflow file, `"3.10"` and `"3.11"` are strings and the numeric comparison won't work for strings.

Reply:
Pyright just fixed this in https://github.com/microsoft/pyright/commit/0a206e884124e20dd31efd1d2a48573ff9d8f6e0

Reply:
Pylance will hopefully resolve this soon